src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -270,40 +362,53 @@ void MulticopterPositionControl::Run()
 					_last_warn = time_stamp_now;
 				}
 
-				failsafe(time_stamp_now, setpoint, _states, !was_in_failsafe);
+				failsafe(time_stamp_now, _setpoint, states, !was_in_failsafe);
 
-				_control.setInputSetpoint(setpoint);
-
-				constraints = {0, NAN, NAN, NAN, NAN, NAN, NAN, NAN, false, {}};
-				_control.setConstraints(constraints);
+				// reset constraints
+				_vehicle_constraints = {0, NAN, NAN, NAN, false, {}};
 
+				_control.setInputSetpoint(_setpoint);
+				_control.setVelocityLimits(_param_mpc_xy_vel_max.get(), _param_mpc_z_vel_max_up.get(), _param_mpc_z_vel_max_dn.get());
 				_control.update(dt);
 			}
 
 			// Publish internal position control setpoints
 			// on top of the input/feed-forward setpoints these containt the PID corrections
 			// This message is used by other modules (such as Landdetector) to determine vehicle intention.
 			vehicle_local_position_setpoint_s local_pos_sp{};
-			local_pos_sp.timestamp = time_stamp_now;
 			_control.getLocalPositionSetpoint(local_pos_sp);
+			local_pos_sp.timestamp = hrt_absolute_time();
 			_local_pos_sp_pub.publish(local_pos_sp);
 
 			// Publish attitude setpoint output
-			// It's important to publish also when disarmed otheriwse the attitude setpoint stays uninitialized.
-			// Not publishing when not running a flight task
-			// in stabilized mode attitude setpoints get ignored
-			// in offboard with attitude setpoints they come from MAVLink directly
 			vehicle_attitude_setpoint_s attitude_setpoint{};
-			attitude_setpoint.timestamp = time_stamp_now;
 			_control.getAttitudeSetpoint(attitude_setpoint);
+			attitude_setpoint.timestamp = hrt_absolute_time();
 			_vehicle_attitude_setpoint_pub.publish(attitude_setpoint);
 
 		} else {
-			// reset the numerical derivatives to not generate d term spikes when coming from non-position controlled operation
-			_vel_x_deriv.reset();
-			_vel_y_deriv.reset();
-			_vel_z_deriv.reset();
```

Comment:
This is still necessary in the else case of `if (_control_mode.flag_control_climb_rate_enabled) {` because the derivative updates https://github.com/PX4/PX4-Autopilot/blob/e00e6f276ed5b4bf885727662cf04710c029011f/src/modules/mc_pos_control/MulticopterPositionControl.cpp#L181-L194 only gets called when climb rate is controlled. Otherwise the derivative is stale during non-climbrate controlled modes and spikes when switching back into a climb rate controlled mode.

Reply:
Actually, that's being updated on every `vehicle_local_position` update now. https://github.com/PX4/PX4-Autopilot/blob/efcbe9705bdf67cf7482609b032cd12f4712d9ec/src/modules/mc_pos_control/MulticopterPositionControl.cpp#L242

I'd still like to do it externally so we have the exact data the controller is using, and then the controller won't need to run at all unless `flag_control_climb_rate_enabled`.

Reply:
Ah true, so for every `vehicle_local_position` there is always a `trajectory_setpoint` (no `_vehicle` 🎉) while the position controller is in use (== `flag_control_climb_rate_enabled`).

Confirmed by:
https://github.com/PX4/PX4-Autopilot/blob/1684a8e0508201bc6f592d7db4e4f1742836462d/src/modules/flight_mode_manager/FlightModeManager.cpp#L101
https://github.com/PX4/PX4-Autopilot/blob/1684a8e0508201bc6f592d7db4e4f1742836462d/src/modules/flight_mode_manager/FlightModeManager.cpp#L145
https://github.com/PX4/PX4-Autopilot/blob/1684a8e0508201bc6f592d7db4e4f1742836462d/src/modules/flight_mode_manager/FlightModeManager.cpp#L521
and MulticopterPositionControl being in the work queue just after FlightModeManager. ✔️